README.md
# js-proper-logger

This repository contains the code for our backend JavaScript logger.

## Using this package

To use this package in a backend JavaScript project:

```
npm install --save @goproperly/js-proper-logger
```

Set an environment variable to set the level that should be logged at
(assuming this is a lambda this can be done via the AWS console):

```sh
PROPERLY_LOG_LEVEL = 20  # Possible values can be found in the `LogLevels` constant
```

And then start logging things:

```javascript
import { ProperLogger } from '@goproperly/js-proper-logger';

// Instances of the logger can be created
const logger = ProperLogger.getLogger('get_offer_details');

// Those instance have methods that allow you to log at different levels of
// severity. Depending on what the `logger.logLevel` is set at (controlled by
// the `PROPERLY_LOG_LEVEL` environment variable) the log may or may not be
// emitted for collection by CloudWatch.
logger.info('Test logging', { exampleTag: 'additional details' });
// This will call `console.log` with `{'message': 'Test logging', 'exampleTag': 'additional details'}`
```

### How does it work

The `ProperLogger` provides us with 3 things:

- An interface for writing structured logs.
- A way to make our structured logs isometric with the logs created by our
  Python backend services.
- A way to change the amount of detail about what is recorded without a new deploy.

The way the `ProperLogger` works is that when a method is called it writes a
message in a JSON format to the `Console`. These messages written to the
console are then collected by `CloudWatch` which we then use to create dashboards
and/or understand what happened when something goes wrong.

#### Metrics

The logger instances can also be used to record metrics:

```javascript
logger.metric('external_call_count', 3);
// This will log to `console.log` and can be used by CloudWatch insights.
```

#### Creating a logger

Instances of the `ProperLogger` should be created using the
`ProperLogger.getLogger` method. This method will return the same instance of
the logger depending on the name provided which allows for reusing the same
logger at different files without passing logger instances between functions.

#### Common tags

This is also a concept of "common tags". These tags are recorded in every log message:

```javascript
logger.addCommonTags({ requestId: 'x-123-fake-request-id' });
logger.info('Enqueued message on event bus');
logger.error('MISSING_VALUE_ERROR', "Expected 'x' to be present on 'Vector'");
// Both of these calls will include the `requestId` in what they write to the Console.

// Common tags can also be cleared
logger.clearCommonTags();
```

#### Log levels

There are four supported levels of logging, which correspond to Python's levels
of logging:

- `logger.debug` (10)
- `logger.info` (20)
- `logger.warning` (30)
- `logger.error` (40)

If you want to decrease the number of logs that the service is creating (for
example if the service was deployed and has been well-behaved for a little
while) you can increase the value of the `PROPERLY_LOG_LEVEL` environment
variable. If the `loggler.logLevel` is greater than the called logging
method's level, nothing will be written to the console (e.g. if
`logger.debug` is called and the `PROPERLY_LOG_LEVEL` is set to
`LogLevels.DEBUG` (10), a message will be written to `console.debug`. If the
`PROPERLY_LOG_LEVEL` log is `LogLevels.INFO` (20) a message will not be written
to `console.debug`).

## Setup

You can install the package by:

```sh
git clone git@github.com:GoProperly/js-proper-logger.git
npm install
```

## Building

This project is built using TypeScript. Because of this there is a "build" step
using Rollup to generate a CommonJS and ESmodule bundles.

```sh
npm run build
```

If you're making many changes you can run the build in "watch" mode:

```sh
npm run dev
```

## Running tests

We use [Jest](https://jestjs.io/docs/en/expect) for testing this library. The
tests run on every pushed commit using [GitHub
actions](https://github.com/GoProperly/js-proper-logger/actions?query=workflow%3A%22Test+%26+Release%22).

Tests can be run with:

```sh
npm run test
npm run test -- --watch
```

## Running linting

This code is linted using `eslint` and autoformatted using `prettier`.
Additionally we make use of a tool called `husky` which will create git hooks
which will check your formatting before allowing you to commit.

```sh
npm run lint
npm run lint:fix
```

## Committing changes

This project adheres to [Semantic Versioning](https://semver.org/) and changes
are documented in the
[CHANGELOG](https://github.com/GoProperly/js-proper-logger/blob/main/CHANGELOG.md).

### Commit messages

Commit messages follow the [Conventional
Commits](https://www.conventionalcommits.org/) format.

You can use `npm run commit` (instead of `git commit`) to help with formatting
your commit message.

## Publishing new versions

New versions are automatically published to the GitHub Package Registry on
pushes to master, using
[`semantic-release`](https://github.com/semantic-release/semantic-release).

Changelog and release notes are generated based on commit messages - no need to
manually update `CHANGELOG.md`.


CHANGELOG.md
### [1.1.1](https://github.com/GoProperly/js-proper-logger/compare/v1.1.0...v1.1.1) (2020-12-31)


### Bug Fixes

* **deps:** [security] bump ini from 1.3.5 to 1.3.8 ([9ee28ab](https://github.com/GoProperly/js-proper-logger/commit/9ee28ab8aebfc5256971abd010f3c727a1e54cf0))

## [1.1.0](https://github.com/GoProperly/js-proper-logger/compare/v1.0.1...v1.1.0) (2020-12-02)


### Features

* add runtime checking to `extraTags` ([6831232](https://github.com/GoProperly/js-proper-logger/commit/6831232974cd9d4698590d0c5ef3c6c83dd66bb9))
* check that `addCommonTags` and the constructor only take valid tags ([7e800e5](https://github.com/GoProperly/js-proper-logger/commit/7e800e59ba601f04b68b69a79f6a7c683de97099))

### [1.0.1](https://github.com/GoProperly/js-proper-logger/compare/v1.0.0...v1.0.1) (2020-09-08)


### Bug Fixes

* update the `ProperLogger.getlogger` to be a singleton factory ([ac23610](https://github.com/GoProperly/js-proper-logger/commit/ac236101b316010d5ca94b5863529057cc22718a))

## 1.0.0 (2020-09-04)


### Features

* deploy first version ([b9fd656](https://github.com/GoProperly/js-proper-logger/commit/b9fd656ef5a5fc217481e7841be4801f6cf61a8d))


test/properLogger.ts
import { ProperLogger } from '..';

describe("LEVELS", () => {
  it("logging levels correspond to acending values as severity increases", () => {
    expect(ProperLogger.LEVELS.DEBUG).toEqual(10);
    expect(ProperLogger.LEVELS.INFO).toEqual(20);
    expect(ProperLogger.LEVELS.INFO).toEqual(30);
    expect(ProperLogger.LEVELS.ERROR).toEqual(40);
  });
});


test/ProperLogger.test.ts
import { ProperLogger, LogLevels } from '..';

describe('LogLevels', () => {
  test('logging levels correspond to acending values as severity increases', () => {
    expect(LogLevels.DEBUG).toEqual(10);
    expect(LogLevels.INFO).toEqual(20);
    expect(LogLevels.WARNING).toEqual(30);
    expect(LogLevels.ERROR).toEqual(40);
  });
});

describe('ProperLogger', () => {
  function setupTestLogger(extraTags?: Record<string, unknown>): ProperLogger {
    const logger = new ProperLogger('test', extraTags);
    logger.console = {
      debug: jest.fn(),
      log: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    };

    return logger;
  }

  describe('getLogger', () => {
    test('returns different instances when given different names', () => {
      const logger1 = ProperLogger.getLogger('outrageous fortune');
      const logger2 = ProperLogger.getLogger('sea of troubles');

      // Note we're using `toBe` rather than `toEqual` not just for the Hamlet
      // pun, but also because we want to check that these are not the same
      // instance, and not just equal to one another.
      expect(logger1).not.toBe(logger2);
    });

    test('returns the same instance when given the same name', () => {
      const logger1 = ProperLogger.getLogger('mortal coil');
      const logger2 = ProperLogger.getLogger('mortal coil');

      // Note we're using `toBe` rather than `toEqual` not just for the Hamlet
      // pun, but also because we want to check that these are the same
      // instance, and not just equal to one another.
      expect(logger1).toBe(logger2);
    });
  });

  describe('debug', () => {
    test('only message', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.DEBUG;

      expect(logger.console.debug).not.toHaveBeenCalled();

      logger.debug('I think there be six Richmonds in the field;');

      expect(logger.console.debug).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'I think there be six Richmonds in the field;',
        })
      );
    });

    test('with `extraTags`', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.DEBUG;

      expect(logger.console.debug).not.toHaveBeenCalled();

      logger.debug('I think there be six Richmonds in the field;', {
        act: 'v',
        scene: 'iv',
      });

      expect(logger.console.debug).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'I think there be six Richmonds in the field;',
          act: 'v',
          scene: 'iv',
        })
      );
    });

    test('with invalid `extraTags`', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.DEBUG;

      expect(logger.console.debug).not.toHaveBeenCalled();

      logger.debug(
        'Rescue, my lord of Norfolk, rescue, rescue!',
        // @ts-expect-error: We're purposefully passing a bad value here
        'this should be an object'
      );

      // We still want to log the message, but we'll drop the tags
      expect(logger.console.debug).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'Rescue, my lord of Norfolk, rescue, rescue!',
        })
      );
      // We log what the tags would have been in a warning.
      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'RECEIVED_INVALID_TAGS',
          warning_message: 'Received invalid value for tags.',
          tags: 'this should be an object',
        })
      );
    });

    test('does not log if `logLevel` is above debug', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.INFO;

      expect(logger.console.debug).not.toHaveBeenCalled();

      logger.debug('Legitimate Edgar, I must have your land.');

      expect(logger.console.debug).not.toHaveBeenCalled();

      // If the log level is then downgraded to `DEBUG` we should then see logs again.
      logger.logLevel = LogLevels.DEBUG;
      logger.debug('Shall top the legitimate. I grow; I prosper.');

      expect(logger.console.debug).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'Shall top the legitimate. I grow; I prosper.',
        })
      );
    });
  });

  describe('info', () => {
    test('only message', () => {
      const logger = setupTestLogger();

      expect(logger.console.log).not.toHaveBeenCalled();

      logger.info('A horse, a horse, my kingdom for a horse!');

      expect(logger.console.log).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'A horse, a horse, my kingdom for a horse!',
        })
      );
    });

    test('with `extraTags`', () => {
      const logger = setupTestLogger();

      expect(logger.console.log).not.toHaveBeenCalled();

      logger.info('A horse, a horse, my kingdom for a horse!', {
        act: 'v',
        scene: 'iv',
      });

      expect(logger.console.log).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'A horse, a horse, my kingdom for a horse!',
          act: 'v',
          scene: 'iv',
        })
      );
    });

    test('with invalid `extraTags`', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.DEBUG;

      expect(logger.console.log).not.toHaveBeenCalled();

      logger.info(
        'His horse is slain, and all on foot he fights,',
        // @ts-expect-error: We're purposefully passing a bad value here
        'this should be an object'
      );

      // We still want to log the message, but we'll drop the tags
      expect(logger.console.log).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'His horse is slain, and all on foot he fights,',
        })
      );
      // We log what the tags would have been in a warning.
      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'RECEIVED_INVALID_TAGS',
          warning_message: 'Received invalid value for tags.',
          tags: 'this should be an object',
        })
      );
    });

    test('does not log if `logLevel` is above INFO', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.WARNING;

      expect(logger.console.log).not.toHaveBeenCalled();

      logger.info('No, I’ll not weep.');

      expect(logger.console.log).not.toHaveBeenCalled();

      // If the log level is then downgraded to `INFO` we should then see logs again.
      logger.logLevel = LogLevels.INFO;
      logger.info('I have full cause of weeping, but this heart ');

      expect(logger.console.log).toHaveBeenCalledWith(
        JSON.stringify({
          message: 'I have full cause of weeping, but this heart ',
        })
      );
    });
  });

  describe('warning', () => {
    test('with only message', () => {
      const logger = setupTestLogger();

      expect(logger.console.warn).not.toHaveBeenCalled();

      logger.warning(
        'Ego_Overload',
        'What do I fear? Myself? There’s none else by.'
      );

      expect(logger.console.warn).toHaveBeenNthCalledWith(
        1,
        JSON.stringify({
          warning_name: 'Ego_Overload',
          warning_message: 'What do I fear? Myself? There’s none else by.',
        })
      );
    });

    test('with `extraTags`', () => {
      const logger = setupTestLogger();

      expect(logger.console.warn).not.toHaveBeenCalled();

      logger.warning(
        'Ego_Overload',
        'What do I fear? Myself? There’s none else by.',
        {
          act: 'v',
          scene: 'iii',
        }
      );

      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'Ego_Overload',
          warning_message: 'What do I fear? Myself? There’s none else by.',
          act: 'v',
          scene: 'iii',
        })
      );
    });

    test('with invalid `extraTags`', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.DEBUG;

      expect(logger.console.warn).not.toHaveBeenCalled();

      logger.warning(
        'KINGDOM_FOR_DISCOUNTED_RATE',
        'A horse, a horse, my kingdom for a horse!',
        // @ts-expect-error: We're purposefully passing a bad value here
        'this should be an object'
      );

      // We still want to log the message, but we'll drop the tags
      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'KINGDOM_FOR_DISCOUNTED_RATE',
          warning_message: 'A horse, a horse, my kingdom for a horse!',
        })
      );
      // We log what the tags would have been in a warning.
      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'RECEIVED_INVALID_TAGS',
          warning_message: 'Received invalid value for tags.',
          tags: 'this should be an object',
        })
      );
    });

    test('does not log if `logLevel` is above WARNING', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.ERROR;

      expect(logger.console.warn).not.toHaveBeenCalled();

      logger.warning(
        'Melodrama',
        'Shall break into a hundred thousand flaws, '
      );

      expect(logger.console.warn).not.toHaveBeenCalled();

      // If the log level is then downgraded to `WARNING` we should then see logs again.
      logger.logLevel = LogLevels.WARNING;
      logger.warning('Melodrama', 'Or ere I’ll weep. O fool, I shall go mad!');

      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'Melodrama',
          warning_message: 'Or ere I’ll weep. O fool, I shall go mad!',
        })
      );
    });
  });

  describe('error', () => {
    test('with only message', () => {
      const logger = setupTestLogger();

      expect(logger.console.error).not.toHaveBeenCalled();

      logger.error(
        'Richard_Confusion',
        'Richard loves Richard; that is, I and I.'
      );

      expect(logger.console.error).toHaveBeenCalledWith(
        JSON.stringify({
          error_name: 'Richard_Confusion',
          error_message: 'Richard loves Richard; that is, I and I.',
        })
      );
    });

    test('with `extraTags`', () => {
      const logger = setupTestLogger();

      expect(logger.console.error).not.toHaveBeenCalled();

      logger.error(
        'Richard_Confusion',
        'Richard loves Richard; that is, I and I.',
        {
          act: 'v',
          scene: 'iii',
        }
      );

      expect(logger.console.error).toHaveBeenCalledWith(
        JSON.stringify({
          error_name: 'Richard_Confusion',
          error_message: 'Richard loves Richard; that is, I and I.',
          act: 'v',
          scene: 'iii',
        })
      );
    });

    test('with invalid `extraTags`', () => {
      const logger = setupTestLogger();
      logger.logLevel = LogLevels.DEBUG;

      expect(logger.console.warn).not.toHaveBeenCalled();

      logger.error(
        'TERRIBLE_PUN',
        'And I will stand the hazard of the die.',
        // @ts-expect-error: We're purposefully passing a bad value here
        'this should be an object'
      );

      // We still want to log the message, but we'll drop the tags
      expect(logger.console.error).toHaveBeenCalledWith(
        JSON.stringify({
          error_name: 'TERRIBLE_PUN',
          error_message: 'And I will stand the hazard of the die.',
        })
      );
      // We log what the tags would have been in a warning.
      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'RECEIVED_INVALID_TAGS',
          warning_message: 'Received invalid value for tags.',
          tags: 'this should be an object',
        })
      );
    });

    test('does not log if `logLevel` is above ERROR', () => {
      const logger = setupTestLogger();
      // We don't have a constant for this. The Python logger has one level
      // above this called `CRITICAL`.
      logger.logLevel = 50;

      expect(logger.console.error).not.toHaveBeenCalled();

      logger.error('Sad_Lear', 'Howl, howl, howl, howl!');

      expect(logger.console.error).not.toHaveBeenCalled();

      // If the log level is then downgraded to `ERROR` we should then see logs again.
      logger.logLevel = LogLevels.ERROR;
      logger.error('Sad_Lear', 'That heaven’s vault should crack');

      expect(logger.console.error).toHaveBeenCalledWith(
        JSON.stringify({
          error_name: 'Sad_Lear',
          error_message: 'That heaven’s vault should crack',
        })
      );
    });
  });

  describe('metric', () => {
    test('can log different kinds of metrics', () => {
      const logger = setupTestLogger();

      logger.metric('num_witches', 3);

      expect(logger.console.log).toHaveBeenNthCalledWith(
        1,
        JSON.stringify({
          metric_name: 'num_witches',
          metric_value: 3,
        })
      );

      logger.metric('kings_name', 'Old Norway');

      expect(logger.console.log).toHaveBeenNthCalledWith(
        2,
        JSON.stringify({
          metric_name: 'kings_name',
          metric_value: 'Old Norway',
        })
      );

      logger.metric('overthinks', true);

      expect(logger.console.log).toHaveBeenNthCalledWith(
        3,
        JSON.stringify({
          metric_name: 'overthinks',
          metric_value: true,
        })
      );

      logger.metric('what_yorick_thinks', null);

      expect(logger.console.log).toHaveBeenNthCalledWith(
        4,
        JSON.stringify({
          metric_name: 'what_yorick_thinks',
          metric_value: null,
        })
      );
    });

    test('allows extraTags', () => {
      const logger = setupTestLogger();

      logger.metric('Cladius', 'mean', { why: 'not a very supportive uncle' });

      expect(logger.console.log).toHaveBeenCalledWith(
        JSON.stringify({
          metric_name: 'Cladius',
          metric_value: 'mean',
          why: 'not a very supportive uncle',
        })
      );
    });

    test('with invalid extraTags', () => {
      const logger = setupTestLogger();

      logger.metric(
        'Polonius',
        'not_good_at_hiding',
        // @ts-expect-error: We're purposefully passing a bad value here
        'this should be an object'
      );

      expect(logger.console.log).toHaveBeenCalledWith(
        JSON.stringify({
          metric_name: 'Polonius',
          metric_value: 'not_good_at_hiding',
        })
      );
      // We log what the tags would have been in a warning.
      expect(logger.console.warn).toHaveBeenCalledWith(
        JSON.stringify({
          warning_name: 'RECEIVED_INVALID_TAGS',
          warning_message: 'Received invalid value for tags.',
          tags: 'this should be an object',
        })
      );
    });

    test('includes common tags', () => {
      const logger = setupTestLogger();

      logger.addCommonTags({ year: 1601 });
      logger.metric('ear poison', 'hebenon');

      expect(logger.console.log).toHaveBeenCalledWith(
        JSON.stringify({
          year: 1601,
          metric_name: 'ear poison',
          metric_value: 'hebenon',
        })
      );
    });
  });

  test('can add common tags', () => {
    const logger = setupTestLogger();

    expect(logger.console.log).not.toHaveBeenCalled();

    logger.addCommonTags({ play: 'Richard III' });
    logger.info(
      'Unless to see my shadow in the sun And descant on mine own deformity',
      {
        act: 'i',
        scene: 'i',
      }
    );

    expect(logger.console.log).toHaveBeenNthCalledWith(
      1,
      JSON.stringify({
        message:
          'Unless to see my shadow in the sun And descant on mine own deformity',
        play: 'Richard III',
        act: 'i',
        scene: 'i',
      })
    );

    logger.addCommonTags({ year: 1633 });
    logger.info('I am determinèd to prove a villain');

    expect(logger.console.log).toHaveBeenNthCalledWith(
      2,
      JSON.stringify({
        message: 'I am determinèd to prove a villain',
        play: 'Richard III',
        year: 1633,
      })
    );
  });

  test('can update common tags', () => {
    const logger = setupTestLogger();

    expect(logger.console.log).not.toHaveBeenCalled();

    logger.addCommonTags({ play: 'Richard III', year: 1633 });
    logger.info('I am determinèd to prove a villain');

    expect(logger.console.log).toHaveBeenNthCalledWith(
      1,
      JSON.stringify({
        message: 'I am determinèd to prove a villain',
        play: 'Richard III',
        year: 1633,
      })
    );

    // This movie version has Ian McKellen
    logger.addCommonTags({ year: 1995 });
    logger.info('I am determinèd to prove a villain');

    expect(logger.console.log).toHaveBeenNthCalledWith(
      2,
      JSON.stringify({
        message: 'I am determinèd to prove a villain',
        play: 'Richard III',
        year: 1995,
      })
    );
  });

  test('will not allow adding invalid common tags', () => {
    const logger = setupTestLogger();

    expect(logger.console.log).not.toHaveBeenCalled();

    // @ts-expect-error: We're purposefully passing a bad value here
    logger.addCommonTags('this should be an object');
    logger.info('I, that am curtailed of this fair proportion', {
      act: 'i',
      scene: 'i',
    });

    expect(logger.console.log).toHaveBeenCalledWith(
      JSON.stringify({
        message: 'I, that am curtailed of this fair proportion',
        act: 'i',
        scene: 'i',
      })
    );
    // We log what the tags would have been in a warning.
    expect(logger.console.warn).toHaveBeenCalledWith(
      JSON.stringify({
        warning_name: 'RECEIVED_INVALID_TAGS',
        warning_message: 'Received invalid value for tags.',
        tags: 'this should be an object',
      })
    );
  });

  test('can clear common tags', () => {
    const logger = setupTestLogger();

    expect(logger.console.log).not.toHaveBeenCalled();

    logger.addCommonTags({ play: 'Richard III', year: 1633 });
    logger.info('some day or two Your Highness shall repose you at the Tower;');

    expect(logger.console.log).toHaveBeenNthCalledWith(
      1,
      JSON.stringify({
        message: 'some day or two Your Highness shall repose you at the Tower;',
        play: 'Richard III',
        year: 1633,
      })
    );

    // This movie version has Ian McKellen
    logger.clearCommonTags();
    logger.info(
      'Then where you please and shall be thought most fit For your best health and recreation.'
    );

    expect(logger.console.log).toHaveBeenNthCalledWith(
      2,
      JSON.stringify({
        message:
          'Then where you please and shall be thought most fit For your best health and recreation.',
      })
    );
  });

  test('can have `commonTags` passed to the constructor', () => {
    const logger = setupTestLogger({ play: 'Much Ado About Nothing' });

    expect(logger.console.log).not.toHaveBeenCalled();

    logger.info('pluck off the bull’s horns and set them in my forehead');

    expect(logger.console.log).toHaveBeenCalledWith(
      JSON.stringify({
        message: 'pluck off the bull’s horns and set them in my forehead',
        play: 'Much Ado About Nothing',
      })
    );
  });

  test('invalid `commonTags` passed to the constructor will be dropped', () => {
    // @ts-expect-error: We're purposefully passing a bad value here
    const logger = setupTestLogger('this should be an object');

    expect(logger.console.log).not.toHaveBeenCalled();

    logger.info('pluck off the bull’s horns and set them in my forehead');

    expect(logger.console.log).toHaveBeenCalledWith(
      JSON.stringify({
        message: 'pluck off the bull’s horns and set them in my forehead',
      })
    );
  });

  test('has `INFO` log level if enviroment variable not set', () => {
    delete process.env.PROPERLY_LOG_LEVEL;

    const logger = setupTestLogger();

    expect(logger.logLevel).toEqual(LogLevels.INFO);
  });

  test('can have log level set by enviroment variable', () => {
    process.env.PROPERLY_LOG_LEVEL = LogLevels.DEBUG.toString();
    expect(setupTestLogger().logLevel).toEqual(LogLevels.DEBUG);

    process.env.PROPERLY_LOG_LEVEL = LogLevels.INFO.toString();
    expect(setupTestLogger().logLevel).toEqual(LogLevels.INFO);
  });

  test('has `INFO` log level if enviroment variable has invalid values', () => {
    process.env.PROPERLY_LOG_LEVEL = undefined;
    expect(setupTestLogger().logLevel).toEqual(LogLevels.INFO);

    process.env.PROPERLY_LOG_LEVEL = '-10';
    expect(setupTestLogger().logLevel).toEqual(LogLevels.INFO);

    process.env.PROPERLY_LOG_LEVEL = 'Potato';
    expect(setupTestLogger().logLevel).toEqual(LogLevels.INFO);
  });
});


src/main.ts
/* eslint-disable no-console, @typescript-eslint/no-explicit-any */
export const LogLevels = {
  DEBUG: 10,
  INFO: 20,
  WARNING: 30,
  ERROR: 40,
};

type Tags = Record<string, any>;

export class ProperLogger {
  name: string;

  commonTags: Tags;

  logLevel: number;

  console: Pick<Console, 'debug' | 'log' | 'warn' | 'error'>;

  static loggerInstances: Record<string, ProperLogger> = {};

  static getLogger(name: string): ProperLogger {
    const memoizedInstance = ProperLogger.loggerInstances[name];
    if (memoizedInstance !== undefined) {
      return memoizedInstance;
    }
    const newInstance = new ProperLogger(name);
    ProperLogger.loggerInstances[name] = newInstance;

    return newInstance;
  }

  constructor(name: string, commonTags?: Tags) {
    this.name = name;
    // Here we grab a reference to the `global.console` so that in tests we can
    // patch this out.
    this.console = global.console;
    this.commonTags = this.checkTags(commonTags || {});

    // Setup the logging level to the value of the `PROPERLY_LOG_LEVEL` or, if
    // the enviroment variable is missing, default to `INFO`.
    this.logLevel = LogLevels.INFO;
    if (process.env.PROPERLY_LOG_LEVEL) {
      const envValue = parseInt(process.env.PROPERLY_LOG_LEVEL, 10);
      if (envValue >= 0) {
        this.logLevel = envValue;
      }
    }
  }

  addCommonTags(commonTags: Tags): void {
    const safeCommonTags = this.checkTags(commonTags);
    this.commonTags = { ...this.commonTags, ...safeCommonTags };
  }

  clearCommonTags(): void {
    this.commonTags = {};
  }

  private prepareLog(message: string | null, extraTags: Tags): string {
    const firstPart: Record<string, string> = {};
    if (message) {
      firstPart.message = message;
    }

    return JSON.stringify({ ...firstPart, ...this.commonTags, ...extraTags });
  }

  // We have seen an issue before where a primative value was passed in as an
  // `extraTags` field. This caused logging to fail which was an undesirable
  // failure mode.
  private checkTags(tags: Tags): Tags {
    if (!(tags && typeof tags === 'object' && !Array.isArray(tags))) {
      this.warning(
        'RECEIVED_INVALID_TAGS',
        'Received invalid value for tags.',
        {
          tags,
        }
      );

      return {};
    }

    return tags;
  }

  debug(message: string, extraTags?: Tags): void {
    if (this.logLevel > LogLevels.DEBUG) {
      return;
    }
    const safeExtraTags = this.checkTags(extraTags || {});
    this.console.debug(this.prepareLog(message, safeExtraTags));
  }

  info(message: string, extraTags?: Tags): void {
    if (this.logLevel > LogLevels.INFO) {
      return;
    }
    const safeExtraTags = this.checkTags(extraTags || {});
    this.console.log(this.prepareLog(message, safeExtraTags));
  }

  warning(name: string, message: string, extraTags?: Tags): void {
    if (this.logLevel > LogLevels.WARNING) {
      return;
    }
    const safeExtraTags = this.checkTags(extraTags || {});
    this.console.warn(
      this.prepareLog(null, {
        warning_name: name,
        warning_message: message,
        ...safeExtraTags,
      })
    );
  }

  error(name: string, message: string, extraTags?: Tags): void {
    if (this.logLevel > LogLevels.ERROR) {
      return;
    }
    const safeExtraTags = this.checkTags(extraTags || {});
    this.console.error(
      this.prepareLog(null, {
        error_name: name,
        error_message: message,
        ...safeExtraTags,
      })
    );
  }

  metric(
    metricName: string,
    metricValue: string | number | boolean | null | undefined,
    extraTags?: Tags
  ): void {
    const safeExtraTags = this.checkTags(extraTags || {});
    this.console.log(
      this.prepareLog(null, {
        metric_name: metricName,
        metric_value: metricValue,
        ...safeExtraTags,
      })
    );
  }
}
